### PR TITLE
[dv] Update chip_sw_i2c_device_tx_rx test to support multiple instances

### LINKED ISSUE #16628: [test] Outstanding tasks for chip_sw_i2c_device_tx_rx
#15700 contains the first version of the chip_sw_i2c_device_tx_rx test.
There are a few outstanding tasks to improve this test:

- [ ] Randomize the I2C instance to use for the test (see e.g. I2C host test).
- [x] Randomize the I2C target masks

### PR DESCRIPTION
**Pull request by vrozic**
Closes #16628
Signed-off-by: Vladimir Rozic <vrozic@lowrisc.org>

### CODE REVIEW

**Review by andreaskurth — COMMENTED** [review]
Thanks @vrozic, this looks functional to me.

One could probably reduce code duplication a bit by assigning variables inside the `switch` statement and passing those variables to DIF calls outside -- but I don't think there's a strong preference for that.

I suggest you add @drewmacrae as reviewer when this is ready

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c [hunk 1]
```diff
@@ -71,6 +71,13 @@ static volatile const uint8_t kI2cByteCount = 0;
 static volatile bool tx_empty_irq_seen = false;
 static volatile bool cmd_complete_irq_seen = false;
 
+/**
+ * these variables store values based on kI2cIdx
+ */
+static uint32_t i2c_irq_fmt_watermark_id;
+static uint32_t i2c_base_addr;
+static top_earlgrey_plic_irq_id_t plic_irqs[9];
```
Comment (vogelpi) [inline sw/device/tests/sim_dv/i2c_device_tx_rx_test.c:79]: `9` should probably be replaced by a constant `kNumI2cIrqs` or similar.

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c [hunk 2]
```diff
@@ -110,47 +116,141 @@ void check_addr(uint8_t addr, dif_i2c_id_t id0, dif_i2c_id_t id1) {
         ((addr & id1.mask) == id1.address));
 }
 
+// handle i2c index related configure
+void config_i2c_with_index(void) {
+  uint8_t i = 0;
+  switch (kI2cIdx) {
+    case 0:
+      i2c_base_addr = TOP_EARLGREY_I2C0_BASE_ADDR;
+      i2c_irq_fmt_watermark_id = kTopEarlgreyPlicIrqIdI2c0FmtWatermark;
```
Comment (vogelpi) [inline sw/device/tests/sim_dv/i2c_device_tx_rx_test.c:125]: Please do the same for the other constants, see also @andreaskurth 's comment.

**Review by vogelpi — COMMENTED** [review]
Thanks @vrozic for the PR, this looks good. I have one minor comment.

In general, I agree with @andreaskurth that code duplication could be reduced by introducing `_id` variables that are assigned the actual constants inside the case statement. The DIF-calls can then use these variables only. This is also what is done for `uart_tx_rx_test.c`, see:
https://github.com/lowRISC/opentitan/blob/2307a48fdcf51ac8460a5194682ef7b719c766f5/sw/device/tests/sim_dv/uart_tx_rx_test.c#L168-L225

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c [hunk 3]
```diff
@@ -71,6 +71,13 @@ static volatile const uint8_t kI2cByteCount = 0;
 static volatile bool tx_empty_irq_seen = false;
 static volatile bool cmd_complete_irq_seen = false;
 
+/**
+ * these variables store values based on kI2cIdx
+ */
+static uint32_t i2c_irq_fmt_watermark_id;
```
Comment (drewmacrae) [inline sw/device/tests/sim_dv/i2c_device_tx_rx_test.c:77]: You don't need another variable to hold this value, main's i2c_ctx has a datamember where it's stored for the test. Because it's an essential part of the i2c_ctx there, the configuration should set it there.

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c [hunk 4]
```diff
@@ -71,6 +71,13 @@ static volatile const uint8_t kI2cByteCount = 0;
...
+static uint32_t i2c_base_addr;
```
Comment (drewmacrae) [inline sw/device/tests/sim_dv/i2c_device_tx_rx_test.c:78]: You don't need a variable to hold this value, the reference i2c points to the base address after it's initialized.

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c (hunk 1 above)
Comment (drewmacrae) [reply sw/device/tests/sim_dv/i2c_device_tx_rx_test.c:79]: It doesn't look like you're reading this table.

**Review by drewmacrae — CHANGES_REQUESTED** [review]
The c code is a bit hard to follow as it's acting a bit unconventionally and isn't properly interacting with the sw structures put in place to help solve these problems.

I think the expected pattern would be to:

- init the i2c to a reference to the appropriate block with a switch statement, so that i2c pointed to the appropriate module,
- init the pinmux with a switch statement so that the i2c had access to the testbench. This part could become a testutil because it should probably be shared between the device and the host test.
- Use generalized offsets so that the configuration can be done in a way that isn't otherwise sensitive to the index.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_i2c_device_tx_rx_vseq.sv [hunk 5]
```diff
@@ -13,7 +13,7 @@ class chip_sw_i2c_device_tx_rx_vseq extends chip_sw_i2c_tx_rx_vseq;
 
   rand int i2c_idx;
   constraint i2c_idx_c {
-    i2c_idx == 0;
+    i2c_idx inside {[0:NUM_I2CS-1]};
```
Comment (jdonjdon) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_i2c_device_tx_rx_vseq.sv:16]: I see you made 3 directed test for i2c_idx and this randomization only be affective when
test runs without using "+i2c_idx=bla".
Can you put a comment in the header of this file as a usage?
Like how to use +i2c_idx and by default it will be randomize index.

Comment (vrozic) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_i2c_device_tx_rx_vseq.sv:21]: No problem, thanks @jdonjdon for reviewing.
I've added the comment with the latest change.
I've implemented directed tests because that's how it was done for the host test, but I think that it would make more sense to have just one test with randomized instance index (the default version), so I'm thinking of changing this.
What do you think?

Comment (jdonjdon) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_i2c_device_tx_rx_vseq.sv:21]: SG, if it is not too much burden, personally, I think one test case with knob is better than 
3 fixed tests.

Comment (vrozic) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_i2c_device_tx_rx_vseq.sv:21]: I agree, I've changed it back to one test

Comment (jdonjdon) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_i2c_device_tx_rx_vseq.sv:21]: LGTM

**Review by jdonjdon — APPROVED** [review]

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c [hunk 6]
```diff
@@ -110,47 +116,141 @@ void check_addr(uint8_t addr, dif_i2c_id_t id0, dif_i2c_id_t id1) {
         ((addr & id1.mask) == id1.address));
 }
 
+// handle i2c index related configure
+void config_i2c_with_index(void) {
```
[the remark this answers is not in the document]
Comment (vrozic) [reply sw/device/tests/sim_dv/i2c_device_tx_rx_test.c]: I agree. Originally I wrote the code this way to mimic what was done on the host side, but now I've reworked it to be more clear. The main changes in the new version are:

- Function `config_i2c_with_index(void)`is removed.
- the base address initialisation is moved to `test_main` and variable `i2c_base_addr` is no longer used
- pinmux initialisation is moved to function `testutils_pinmux(const dif_pinmux_t *pinmux, uint8_t kI2cIdx);` which was added to `i2c_testutils.c`
- function `config_irqs_with_index(top_earlgrey_plic_irq_id_t *plic_irqs)` was added to `test_main` to setup the interrupt request.

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c (hunk 4 above)
Comment (vrozic) [reply sw/device/tests/sim_dv/i2c_device_tx_rx_test.c]: I agree. This variable was removed in the latest version

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c (hunk 3 above)
Comment (vrozic) [reply sw/device/tests/sim_dv/i2c_device_tx_rx_test.c]: I've moved watermark ID initialisation to function `void ottf_external_isr(void) `.
This function now uses a local variable `i2c_irq_fmt_watermark_id` in order to avoid code duplication.

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c (hunk 1 above)
Comment (vrozic) [reply sw/device/tests/sim_dv/i2c_device_tx_rx_test.c]: @vogelpi OK, I've handled this by defining a constant `#define NUM_I2C_IRQS 6`

@drewmacrae Indeed, that was an ommision in the previous version. `plic_irqs[]`is now used in the `test_main` function.

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c (hunk 2 above)
Comment (vrozic) [reply sw/device/tests/sim_dv/i2c_device_tx_rx_test.c]: This is changed in the latest version.
Most of these constants are now in the function `i2c_testutils_pinmux` in `12c_testutils.c`

**Comment by vogelpi** [comment]
Hey @drewmacrae ,
would you mind having another look at this PR please? Since you've requested changes initially, we are not able to merge this without your explicit approval. Thanks!

File: sw/device/lib/testing/i2c_testutils.c [hunk 7]
```diff
@@ -164,3 +166,56 @@ bool i2c_testutils_target_check_wr(dif_i2c_t *i2c, uint8_t byte_count,
 
   return i2c_testutils_target_check_end(i2c, cont_byte);
 }
+
+void i2c_testutils_pinmux(const dif_pinmux_t *pinmux, uint8_t kI2cIdx) {
```
Comment (drewmacrae) [inline sw/device/lib/testing/i2c_testutils.c:170]: Please rename this so it's consistent with the naming scheme and has a verb.
Maybe: "i2c_testutils_connect_i2c_to_pinmux_pins"

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c [hunk 8]
```diff
@@ -110,47 +134,89 @@ void check_addr(uint8_t addr, dif_i2c_id_t id0, dif_i2c_id_t id1) {
...
+
+    // Assign a default priority
+    CHECK_DIF_OK(dif_rv_plic_irq_set_priority(&plic, plic_irqs[i],
+                                              kDifRvPlicMaxPriority));
+  }
+
+  switch (kI2cIdx) {
+    case 0:
+      CHECK_DIF_OK(dif_i2c_init(
+          mmio_region_from_addr(TOP_EARLGREY_I2C0_BASE_ADDR), &i2c));
```
Comment (drewmacrae) [inline sw/device/tests/sim_dv/i2c_device_tx_rx_test.c:202]: The dif functions all use the i2c struct to pass a reference to the i2c so I'm alarmed to see it being set so late here. They might enable much more concise expressions of intent.

For example, if you use the difs you could operate on [dif_i2c_irq_t](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/dif/autogen/dif_i2c_autogen.h;bpv=1;bpt=1;l=149?gsn=dif_i2c_irq_t&gs=kythe%3A%2F%2Fgithub.com%2FlowRISC%2Fopentitan%3Flang%3Dc%252B%252B%23xscMge-OWf93wHtGpa7YEgTb3bAM2SRabz5t3uwBMAE) instead of uints and write interrupt handling code that just uses the &i2c reference to work on the correct instance.

**Review by drewmacrae — APPROVED** [review]
I've added a number of comments that could be addressed to make this a better example of how we intend the DIFs to facilitate writing these sorts of programs. It looks like the host test is similarly structured so this could be done in a followup.

File: sw/device/lib/testing/i2c_testutils.c [hunk 9]
```diff
@@ -164,3 +166,56 @@ bool i2c_testutils_target_check_wr(dif_i2c_t *i2c, uint8_t byte_count,
...
+      i2c_pinmux_in1_id = kTopEarlgreyPinmuxInselIob11;
+      i2c_pinmux_in2_id = kTopEarlgreyPinmuxInselIob12;
+      i2c_pinmux_insel_scl_id = kTopEarlgreyPinmuxPeripheralInI2c2Scl;
+      i2c_pinmux_insel_sda_id = kTopEarlgreyPinmuxPeripheralInI2c2Sda;
+      i2c_pinmux_outsel_scl_id = kTopEarlgreyPinmuxOutselI2c2Scl;
+      i2c_pinmux_outsel_sda_id = kTopEarlgreyPinmuxOutselI2c2Sda;
+      break;
+    default:
+      LOG_FATAL("Unsupported i2c index %d", kI2cIdx);
+  }
```
Comment (alphan) [inline sw/device/lib/testing/i2c_testutils.c:211]: this should probably be an array of structs defined as a static constant in this file.

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c [hunk 10]
```diff
@@ -71,20 +70,45 @@ static volatile const uint8_t kI2cByteCount = 0;
 static volatile bool tx_empty_irq_seen = false;
 static volatile bool cmd_complete_irq_seen = false;
 
+/**
+ * This constant indicates the number of interrupt requests.
+ */
+#define NUM_I2C_IRQS 6
```
Comment (alphan) [inline sw/device/tests/sim_dv/i2c_device_tx_rx_test.c:76]: Please use a anonymous enum instead of a macro for consistency, i.e.:

```
enum {
   kFoo = 6,
};
```

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c [hunk 11]
```diff
@@ -110,47 +134,89 @@ void check_addr(uint8_t addr, dif_i2c_id_t id0, dif_i2c_id_t id1) {
         ((addr & id1.mask) == id1.address));
 }
 
+void config_irqs_with_index(top_earlgrey_plic_irq_id_t *plic_irqs) {
+  uint8_t i = 0;
```
Comment (alphan) [inline sw/device/tests/sim_dv/i2c_device_tx_rx_test.c:138]: nit: size_t

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c [hunk 12]
```diff
@@ -110,47 +134,89 @@ void check_addr(uint8_t addr, dif_i2c_id_t id0, dif_i2c_id_t id1) {
...
+    case 2:
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2CmdComplete;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2TxStretch;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2TxOverflow;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2AcqFull;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2UnexpStop;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2HostTimeout;
+      break;
+    default:
+      LOG_FATAL("Unsupported i2c index %d", kI2cIdx);
```
Comment (alphan) [inline sw/device/tests/sim_dv/i2c_device_tx_rx_test.c:165]: nit: this could also be simplified using an array of structs.

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c [hunk 13]
```diff
@@ -110,47 +134,89 @@ void check_addr(uint8_t addr, dif_i2c_id_t id0, dif_i2c_id_t id1) {
...
+  for (uint32_t i = 0; i < ARRAYSIZE(plic_irqs); ++i) {
+    CHECK_DIF_OK(dif_rv_plic_irq_set_enabled(
+        &plic, plic_irqs[i], kTopEarlgreyPlicTargetIbex0, kDifToggleEnabled));
+
+    // Assign a default priority
+    CHECK_DIF_OK(dif_rv_plic_irq_set_priority(&plic, plic_irqs[i],
+                                              kDifRvPlicMaxPriority));
+  }
+
+  switch (kI2cIdx) {
```
Comment (alphan) [inline sw/device/tests/sim_dv/i2c_device_tx_rx_test.c:199]: this switch statement too could be simplified using an array of structs.

I'd suggest defining a struct that contains all the values that these 3 switch statements need and define an array with three elements, one per i2c.

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c (hunk 10 above)
Comment (vrozic) [reply sw/device/tests/sim_dv/i2c_device_tx_rx_test.c]: Thanks. I've fixed it now.

File: sw/device/tests/sim_dv/i2c_device_tx_rx_test.c (hunk 11 above)
Comment (vrozic) [reply sw/device/tests/sim_dv/i2c_device_tx_rx_test.c]: fixed

File: sw/device/lib/testing/i2c_testutils.c (hunk 9 above)
Comment (vrozic) [reply sw/device/lib/testing/i2c_testutils.c]: I agree, thanks for the suggestion. It is fixed now.

**Comment by vogelpi** [comment]
CI has finally passed, I am thus merging this. Thanks @vrozic for the PR and @andreaskurth , @jdonjdon , @alphan and @drewmacrae for your reviews!
